plugins/vpp/l3plugin/descriptor/dhcp_proxy.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2019 PANTHEON.tech
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package descriptor
+
+import (
+	"net"
+
+	"github.com/ligato/cn-infra/logging"
+	"github.com/ligato/cn-infra/utils/addrs"
+	l3 "github.com/ligato/vpp-agent/api/models/vpp/l3"
+	kvs "github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	"github.com/ligato/vpp-agent/plugins/vpp/l3plugin/descriptor/adapter"
+	"github.com/ligato/vpp-agent/plugins/vpp/l3plugin/vppcalls"
+	"github.com/pkg/errors"
+)
+
+const (
+	// ProxyArpInterfaceDescriptorName is the name of the descriptor.
+	DHCPProxyDescriptorName = "dhcp-proxy"
+
+	//dependecy labels
+	vrfTableDependency = "vrf-table-exists"
+)
+
+// DHCPProxyDescriptor teaches KVScheduler how to configure VPP DHCP proxy.
+type DHCPProxyDescriptor struct {
+	log              logging.Logger
+	dhcpProxyHandler vppcalls.DHCPProxyAPI
+	scheduler        kvs.KVScheduler
+}
+
+// NewDHCPProxyDescriptor creates a new instance of the DHCPProxyDescriptor.
+func NewDHCPProxyDescriptor(scheduler kvs.KVScheduler,
+	dhcpProxyHandler vppcalls.DHCPProxyAPI, log logging.PluginLogger) *kvs.KVDescriptor {
+
+	ctx := &DHCPProxyDescriptor{
+		scheduler:        scheduler,
+		dhcpProxyHandler: dhcpProxyHandler,
+		log:              log.NewLogger("dhcp-proxy-descriptor"),
+	}
+
+	typedDescr := &adapter.DHCPProxyDescriptor{
+		Name:          DHCPProxyDescriptorName,
+		KeySelector:   l3.ModelDHCPProxy.IsKeyValid,
+		KeyLabel:      l3.ModelDHCPProxy.StripKeyPrefix,
+		NBKeyPrefix:   l3.ModelDHCPProxy.KeyPrefix(),
+		ValueTypeName: l3.ModelDHCPProxy.ProtoName(),
+		Create:        ctx.Create,
+		Delete:        ctx.Delete,
+		Retrieve:      ctx.Retrieve,
+		Dependencies:  ctx.Dependencies,
+		Validate:      ctx.Validate,
+	}
+	return adapter.NewDHCPProxyDescriptor(typedDescr)
+}
+
+func (d *DHCPProxyDescriptor) Validate(key string, value *l3.DHCPProxy) error {
+
+	ipAddr := net.ParseIP(value.SourceIpAddress)
+	if ipAddr == nil {
+		return errors.Errorf("invalid IP address: %q", value.SourceIpAddress)
+	}
+
+	isIPv4 := true
+	if ipAddr.To4() == nil {
+		isIPv4 = false
+	}
+
+	for _, server := range value.Servers {
+		serverIpAddr := net.ParseIP(server.IpAddress)
+		if serverIpAddr == nil {
+			return errors.Errorf("invalid IP address: %q", server.IpAddress)
+		}
+
+		if isIPv4 {
+			if serverIpAddr.To4() == nil {
+				return errors.Errorf("Server address must be IPv4 IP address: %q", server.IpAddress)
```

Comment:
I think the error message might be confusing in log.. it should mention the protocol mismatch with the source IP. 

Something like: `server address must use same IP protocol as source address (IPv4)` is probably more clear.